File: backend/src/db/models/instrumentInfo.py
from db.models.base import BaseModelWithUndefined


class InstrumentInfoData(BaseModelWithUndefined):
    instrument: str
    priceStep: float
    quantityStep: float


class InstrumentInfo(BaseModelWithUndefined):
    id: int
    instrument: str
    priceStep: float
    quantityStep: float

File: frontend/src/common/instrumentInfo.ts
import { InstrumentInfo, TransactionData } from '../api/backend/types';
import { warnDefined } from './assert';
import { getDecimalsStep } from './value';

export const findInstrumentInfo = (
  infos: InstrumentInfo[],
  transaction: TransactionData<'grid'>,
): InstrumentInfo =>
  warnDefined(
    infos.find((info) => info.instrument === transaction.instrument),
    {
      id: -1,
      instrument: transaction.instrument,
      priceStep: Math.min(
        getDecimalsStep(transaction.minPrice),
        getDecimalsStep(transaction.maxPrice),
        getDecimalsStep(transaction.startPrice),
        getDecimalsStep(transaction.endPrice),
      ),
      quantityStep: getDecimalsStep(transaction.quantity),
    },
    `Instrument info not found: ${transaction.instrument}`,
  );
